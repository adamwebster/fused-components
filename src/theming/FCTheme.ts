import React from 'react';

export interface Theme {
    theme?: string,
    children?: any,
}

export const FCTheme = React.createContext<Theme | null>(null);

export const FCThemeProvider = FCTheme.Provider;

export const FCThemeConsumer = FCTheme.Consumer;