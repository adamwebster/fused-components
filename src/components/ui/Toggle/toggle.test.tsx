import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';
import 'jest-styled-components';
import { Toggle } from './index';
import { FCThemeProvider } from '../../../theming/FCTheme';

afterEach(cleanup);

describe('Toggle Tests', () => {
  test('It renders correctly', () => {
    render(<Toggle />);
  });
  test('It is not turned on when first loaded', () => {
    const { getByRole } = render(<Toggle />);
    const toggle = getByRole('checkbox');
    expect(toggle.getAttribute('aria-checked')).toBe('false');
  });
  test('It is not turned on when active is set to false', () => {
    const { getByRole } = render(<Toggle active={false} />);
    const toggle = getByRole('checkbox');
    expect(toggle.getAttribute('aria-checked')).toBe('false');
  });
  test('It renders when set to active', () => {
    const { getByRole } = render(<Toggle active />);
    const toggle = getByRole('checkbox');
    expect(toggle.getAttribute('aria-checked')).toBe('true');
  });
  test('When showLabels is on the off label show', () => {
    const { getByText } = render(<Toggle showLabels />);
    expect(getByText('Off')).toBeInTheDocument();
  });

  test('When showLabels is on and it is active the on label show', () => {
    const { getByText } = render(<Toggle active showLabels />);
    expect(getByText('On')).toBeInTheDocument();
  });
  test('Clicking the toggle activates the toggle', () => {
    let active = false;
    const click = jest.fn(() => {
      active = true;
    });
    const { getByRole, rerender } = render(<Toggle onClick={click} active={active} />);
    const toggle = getByRole('checkbox');
    expect(toggle.getAttribute('aria-checked')).toBe('false');
    fireEvent.click(toggle);
    expect(click).toHaveBeenCalledTimes(1);
    rerender(<Toggle onClick={click} active={active} />);
    expect(toggle.getAttribute('aria-checked')).toBe('true');
  });

  test('Pressing enter on the toggle activates the toggle', () => {
    let active = false;
    const click = jest.fn(() => {
      active = true;
    });
    const { getByRole, rerender } = render(<Toggle onClick={click} active={active} />);
    const toggle = getByRole('checkbox');
    expect(toggle.getAttribute('aria-checked')).toBe('false');
    fireEvent.keyDown(toggle, { key: 'Enter' });
    expect(click).toHaveBeenCalledTimes(1);
    rerender(<Toggle onClick={click} active={active} />);
    expect(toggle.getAttribute('aria-checked')).toBe('true');
  });

  test('Pressing space on the toggle activates the toggle', () => {
    let active = false;
    const click = jest.fn(() => {
      active = true;
    });
    const { getByRole, rerender } = render(<Toggle onClick={click} active={active} />);
    const toggle = getByRole('checkbox');
    expect(toggle.getAttribute('aria-checked')).toBe('false');
    fireEvent.keyDown(toggle, { key: ' ' });
    expect(click).toHaveBeenCalledTimes(1);
    rerender(<Toggle onClick={click} active={active} />);
    expect(toggle.getAttribute('aria-checked')).toBe('true');
  });

  test('Pressing any other key on the toggle does not activates the toggle', () => {
    let active = false;
    const click = jest.fn(() => {
      active = true;
    });
    const { getByRole, rerender } = render(<Toggle onClick={click} active={active} />);
    const toggle = getByRole('checkbox');
    expect(toggle.getAttribute('aria-checked')).toBe('false');
    fireEvent.keyDown(toggle, { key: 'A' });
    expect(click).toHaveBeenCalledTimes(0);
    rerender(<Toggle onClick={click} active={active} />);
    expect(toggle.getAttribute('aria-checked')).toBe('false');
  });

  test('It renders as expected in dark mode', () => {
    const { container } = render(
      <FCThemeProvider value={{ theme: 'dark' }}>
        <Toggle showLabels active={false} />
      </FCThemeProvider>,
    );
    expect(container).toMatchSnapshot();
  });

  test('It renders as expected in dark mode when active', () => {
    const { container } = render(
      <FCThemeProvider value={{ theme: 'dark' }}>
        <Toggle showLabels active={true} />
      </FCThemeProvider>,
    );
    expect(container).toMatchSnapshot();
  });
});
