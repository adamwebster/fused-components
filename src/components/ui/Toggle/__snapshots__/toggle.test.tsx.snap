// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle Tests It renders as expected in dark mode 1`] = `
.c0 {
  width: 40px;
  height: 20px;
  border-radius: 15px;
  padding: 3px;
  background-color: #486581;
  position: relative;
  cursor: pointer;
  border: solid 1px rgba(0,0,0,.1);
}

.c1 {
  width: 20px;
  height: 20px;
  background-color: #bcccdc;
  position: absolute;
  box-sizing: border-box;
  box-shadow: 0 0 5px rgba(0,0,0,0.25);
  border: solid 1px rgba(0,0,0,.1);
  border-radius: 15px;
  -webkit-transition: all 0.1s ease;
  transition: all 0.1s ease;
  left: 3px;
}

.c2 {
  color: #999999;
  font-size: 10px;
  box-sizing: border-box;
}

.c2:first-of-type {
  color: #102A43;
  padding-left: 5px;
  padding-right: 4px;
}

.c3 {
  color: #fff;
  font-size: 10px;
  box-sizing: border-box;
}

.c3:first-of-type {
  color: #102A43;
  padding-left: 5px;
  padding-right: 4px;
}

<div>
  <div
    aria-checked="false"
    class="c0"
    role="checkbox"
  >
    <div
      class="c1"
    />
    <span
      class="c2"
    >
      On
    </span>
    <span
      class="c3"
    >
      Off
    </span>
  </div>
</div>
`;

exports[`Toggle Tests It renders as expected in dark mode when active 1`] = `
.c0 {
  width: 40px;
  height: 20px;
  border-radius: 15px;
  padding: 3px;
  background-color: #829ab1;
  position: relative;
  cursor: pointer;
  border: solid 1px rgba(0,0,0,.1);
}

.c1 {
  width: 20px;
  height: 20px;
  background-color: #bcccdc;
  position: absolute;
  box-sizing: border-box;
  box-shadow: 0 0 5px rgba(0,0,0,0.25);
  border: solid 1px rgba(0,0,0,.1);
  border-radius: 15px;
  -webkit-transition: all 0.1s ease;
  transition: all 0.1s ease;
  left: 23px;
}

.c2 {
  color: #999999;
  font-size: 10px;
  box-sizing: border-box;
}

.c2:first-of-type {
  color: #102A43;
  padding-left: 5px;
  padding-right: 4px;
}

.c3 {
  color: #fff;
  font-size: 10px;
  box-sizing: border-box;
}

.c3:first-of-type {
  color: #102A43;
  padding-left: 5px;
  padding-right: 4px;
}

<div>
  <div
    aria-checked="true"
    class="c0"
    role="checkbox"
  >
    <div
      class="c1"
    />
    <span
      class="c2"
    >
      On
    </span>
    <span
      class="c3"
    >
      Off
    </span>
  </div>
</div>
`;
